app/static/app/js/components/TaskListItem.jsx

```diff
@@ -0,0 +1,260 @@
+    if (this.state.task.processing_time) this.setState({time: this.state.task.processing_time});
+  }
+
+  componentDidMount(){
+    if (this.shouldRefresh()) this.refreshTimeout = setTimeout(() => this.refresh(), this.props.refreshInterval || 3000);
+
+    // Load timer if we are in running state
+    if (this.state.task.status === statusCodes.RUNNING) this.loadTimer(this.state.task.processing_time);
+  }
+
+  refresh(){
+    // Fetch
+    this.refreshRequest = $.getJSON(`/api/projects/${this.state.task.project}/tasks/${this.state.task.id}/`, json => {
+      if (json.id){
+        let oldStatus = this.state.task.status;
+
+        this.setState({task: json, actionButtonsDisabled: false});
+
+        // Update timer if we switched to running
+        if (oldStatus !== this.state.task.status){
+          if (this.state.task.status === statusCodes.RUNNING){
+            this.loadTimer(this.state.task.processing_time);
+          }else{
+            this.setState({time: this.state.task.processing_time});
+            this.unloadTimer();
+          }
+        }
+      }else{
+        console.warn("Cannot refresh task: " + json);
+      }
+    })
+    .always((_, textStatus) => {
+      if (textStatus !== "abort"){
+        if (this.shouldRefresh()) this.refreshTimeout = setTimeout(() => this.refresh(), this.props.refreshInterval || 3000);
+      }
+    });
+  }
+
+  componentWillUnmount(){
+    this.unloadTimer();
+    if (this.refreshTimeout) clearTimeout(this.refreshTimeout);
+    if (this.refreshRequest) this.refreshRequest.abort();
+  }
+
+  toggleExpanded(){
+    this.setState({
+      expanded: !this.state.expanded
+    });
+  }
+
+  consoleOutputUrl(line){
+    return `/api/projects/${this.state.task.project}/tasks/${this.state.task.id}/output/?line=${line}`;
+  }
+
+  hoursMinutesSecs(t){
+    if (t === 0 || t === -1) return "-- : -- : --";
+
+    let ch = 60 * 60 * 1000,
+          cm = 60 * 1000,
+          h = Math.floor(t / ch),
+          m = Math.floor( (t - h * ch) / cm),
+          s = Math.round( (t - h * ch - m * cm) / 1000),
+          pad = function(n){ return n < 10 ? '0' + n : n; };
+    if( s === 60 ){
+      m++;
+      s = 0;
+    }
+    if( m === 60 ){
+      h++;
+      m = 0;
+    }
+    return [pad(h), pad(m), pad(s)].join(':');
+  }
+
+  render() {
+    let name = this.state.task.name !== null ? this.state.task.name : `Task #${this.state.task.id}`;
+    
+    let status = statusCodes.description(this.state.task.status);
+    if (status === "") status = "Uploading images";
+    if (this.state.task.pending_action !== null) status = pendingActions.description(this.state.task.pending_action);
+
+    let expanded = "";
+    if (this.state.expanded){
+      let actionButtons = [];
+      const addActionButton = (label, className, icon, onClick) => {
+        actionButtons.push({
+          className, icon, label, onClick
+        });
+      };
+      const genActionApiCall = (action) => {
```

Comment:
Move this out of `render()` (into `componentDidMount`?) to avoid creating lots of closures that do the same thing?